File: pk-hotel-frontend/src/Hotel/Hotel.ts
import dayjs from 'dayjs';

export type HotelDTO = {
  name: string;
  owner: string;
  registerDate: string;
  country: string;
  city: string;
  address: string;
}

export class Hotel {
  //private properties
  private _name: string;
  private _owner: string;
  private _registerDate: string;
  private _country: string;
  private _city: string;
  private _address: string;

  constructor(initializer: Omit<Hotel, 'registerDate'> & Partial<Pick<Hotel, 'registerDate'>>){
    this._name = initializer.name;
    this._owner = initializer.owner;
    this._registerDate = dayjs().format('YYYY-MM-DD');
    this._country = initializer.country;
    this._city = initializer.city;
    this._address = initializer.address;
  }

  public get name(): string {
    return this._name;
  }

  public set name(name: string){
    this._name = name;
  }

  public get owner(): string {
    return this._owner;
  }

  public set owner(owner: string){
    this._owner = owner;
  }

  public get registerDate(): string {
    return this._registerDate;
  }

  public set registerDate(registerDate: string){
    this._registerDate = registerDate;
  }

  public get country(): string {
    return this._country;
  }

  public set country(country: string){
    this._country = country;
  }

  public get city(): string {
    return this._city;
  }

  public set city(city: string){
    this._city = city;
  }

  public get address(): string {
    return this._address;
  }

  public set address(address: string){
    this._address = address;
  }
}

export const transfromHotel = (hotel: Hotel): HotelDTO => {
  return {
    name: hotel.name,
    owner: hotel.owner,
    registerDate: hotel.registerDate,
    country: hotel.country,
    city: hotel.city,
    address: hotel.address
  };
}

File: pk-hotel-backend/src/main/java/edu/zespol5/pkhotelbackend/model/hotel/Hotel.java
package edu.zespol5.pkhotelbackend.model.hotel;

import edu.zespol5.pkhotelbackend.model.room.Room;
import jakarta.persistence.*;
import lombok.Data;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;
import java.util.HashSet;
import java.util.Set;

/**
 * Represents a hotel entity in the system.
 * <p>
 * A hotel has an associated set of rooms and basic details like name, owner, registration date,
 * and address. This class is mapped to a database table called "hotel".
 * </p>
 *
 * @see Room
 */
@Getter
@Setter
@Entity
@Table(name = "hotel")
public class Hotel {

    /**
     * The unique identifier of the hotel.
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    /**
     * The name of the hotel.
     */
    private String name;

    /**
     * The owner of the hotel.
     */
    private String owner;

    /**
     * The registration date of the hotel.
     */
    private LocalDate registerDate;

    /**
     * The country where the hotel is located.
     */
    private String country;

    /**
     * The city where the hotel is located.
     */
    private String city;

    /**
     * The address of the hotel.
     */
    private String address;

    /**
     * The set of rooms associated with this hotel.
     * <p>
     * This is a one-to-many relationship with the {@link Room} entity, and rooms are automatically
     * deleted if the hotel is removed.
     * </p>
     */
    @OneToMany(mappedBy = "hotel", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<Room> rooms = new HashSet<>();

    /**
     * Default constructor for Hotel.
     */
    public Hotel() {
        // Default constructor
    }

    /**
     * Adds a room to the hotel.
     *
     * @param room the room to be added to the hotel
     */
    public void addRoom(Room room) {
        rooms.add(room);
    }

    /**
     * Removes a room from the hotel.
     *
     * @param room the room to be removed from the hotel
     */
    public void removeRoom(Room room) {
        rooms.remove(room);
    }
}
